File: platform_src/android/app/src/main/java/org/elastos/trinity/runtime/AppManager.java
 /*
  * Copyright (c) 2018 Elastos Foundation
  *
  * Permission is hereby granted, free of charge, to any person obtaining a copy
  * of this software and associated documentation files (the "Software"), to deal
  * in the Software without restriction, including without limitation the rights
  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  * copies of the Software, and to permit persons to whom the Software is
  * furnished to do so, subject to the following conditions:
  *
  * The above copyright notice and this permission notice shall be included in all
  * copies or substantial portions of the Software.
  *
  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */

package org.elastos.trinity.runtime;

import android.app.AlertDialog;
//import android.app.FragmentManager;
//import android.app.FragmentTransaction;
import android.content.DialogInterface;
import android.content.res.AssetManager;
import android.content.res.Configuration;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.view.View;

import org.apache.cordova.PluginManager;
import org.elastos.trinity.runtime.R;
import org.json.JSONException;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;

public class AppManager {

    /** The internal message */
    public static final int MSG_TYPE_INTERNAL = 1;
    /** The internal return message. */
    public static final int MSG_TYPE_IN_RETURN = 2;
    /** The external launcher message */
    public static final int MSG_TYPE_EXTERNAL_LAUNCHER = 3;
    /** The external install message */
    public static final int MSG_TYPE_EXTERNAL_INSTALL = 4;
    /** The external return message. */
    public static final int MSG_TYPE_EX_RETURN = 5;

    private static AppManager appManager;
    public WebViewActivity activity;
    public WebViewFragment curFragment = null;
    ManagerDBAdapter dbAdapter = null;

    public String appsPath = null;
    public String dataPath = null;
    public String tempPath = null;

    private AppInstaller installer;

    protected LinkedHashMap<String, AppInfo> appInfos;
    private ArrayList<String> lastList = new ArrayList<String>();
    public AppInfo[] appList;

    private AppInfo launcherInfo;

    private ArrayList<String>  installUriList = new ArrayList<String>();
    private boolean launcherReady = false;

    AppManager(WebViewActivity activity) {
        AppManager.appManager = this;
        this.activity = activity;

        appsPath = activity.getFilesDir() + "/apps/";
        dataPath = activity.getFilesDir() + "/data/";
        tempPath = activity.getFilesDir() + "/temp/";

        Boolean first = false;
        File destDir = new File(appsPath);
        if (!destDir.exists()) {
            destDir.mkdirs();
            first = true;
        }
        destDir = new File(dataPath);
        if (!destDir.exists()) {
            destDir.mkdirs();
        }
        destDir = new File(tempPath);
        if (!destDir.exists()) {
            destDir.mkdirs();
        }

        destDir = new File(dataPath + "launcher");
        if (!destDir.exists()) {
            destDir.mkdirs();
        }

        dbAdapter = new ManagerDBAdapter(activity);
//        dbAdapter.clean();

        installer = new AppInstaller();
        installer.init(activity, dbAdapter, appsPath, dataPath, tempPath);

        if (first) {
            saveLauncherInfo();
        }

        appList = dbAdapter.getAppInfos();
        saveBuiltInAppInfos();

        refreashInfos();
    }

    public static AppManager getShareInstance() {
        return AppManager.appManager;
    }

    private void saveLauncherInfo() {
        AssetManager manager = activity.getAssets();
        try {
            InputStream input = manager.open("www/launcher/manifest.json");
            AppInfo info = installer.parseManifest(input, 1);
            installer.copyAssetsFolder("www/launcher", appsPath + info.app_id);
            info.built_in = 1;
            dbAdapter.addAppInfo(info);

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public AppInfo getLauncherInfo() {
        if (launcherInfo == null) {
            launcherInfo = dbAdapter.getLauncherInfo();
        }
        return launcherInfo;
    }

    private void refreashInfos() {
        appList = dbAdapter.getAppInfos();
        appInfos = new LinkedHashMap();
        for (int i = 0; i < appList.length; i++) {
            appInfos.put(appList[i].app_id, appList[i]);
        }
    }

    public AppInfo getAppInfo(String id) {
        if (id == "launcher") {
            return getLauncherInfo();
        }
        else {
            return appInfos.get(id);
        }
    }

    public HashMap<String, AppInfo> getAppInfos() {
        return appInfos;
    }

    public String getStartPath(AppInfo info) {
        if (info.remote == 0) {
            return getAppUrl(info) + info.start_url;
        }
        else {
            return info.start_url;
        }
    }

    public String getAppPath(AppInfo info) {
        if (info.remote == 0) {
            return appsPath + info.app_id + "/";
        }
        else {
            return info.start_url.substring(0, info.start_url.lastIndexOf("/") + 1);
        }
    }

    public String getAppUrl(AppInfo info) {
        String url = getAppPath(info);
        if (info.remote == 0) {
            url = "file://" + url;
        }
        return url;
    }

    public String getDataPath(String id) {
        if (id == "launcher") {
            id = getLauncherInfo().app_id;
        }
        return dataPath + id + "/";
    }

    public String getDataUrl(String id) {
        return "file://" + getDataPath(id);
    }


    public String getTempPath(String id) {
        if (id == "launcher") {
            id = getLauncherInfo().app_id;
        }
        return tempPath + id + "/";
    }

    public String getTempUrl(String id) {
        return "file://" + getTempPath(id);
    }

    public String resetPath(String dir, String origin) {
        if (origin.indexOf("http://") != 0 && origin.indexOf("https://") != 0
                && origin.indexOf("file:///") != 0) {
            while (origin.startsWith("/")) {
                origin = origin.substring(1);
            }
            origin = dir + origin;
        }
        return origin;
    }

    public void saveBuiltInAppInfos(){
        AssetManager manager = activity.getAssets();
        try {
            String[] appdirs= manager.list("www/built-in");
            for (String appdir : appdirs) {
                boolean needInstall = true;
                for (int i = 0; i < appList.length; i++) {
                    if (appdir.equals(appList[i].app_id)) {
                        needInstall = false;
                        break;
                    }
                }

                if (needInstall) {
                    installer.copyAssetsFolder("www/built-in/" + appdir, appsPath + appdir);
                    InputStream input = new FileInputStream(appsPath + appdir + "/manifest.json");
                    AppInfo info = installer.parseManifest(input, 0);

                    info.built_in = 1;
                    dbAdapter.addAppInfo(info);
                }
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public AppInfo install(String url) throws Exception  {
        AppInfo info = installer.install(this, url);
        if (info != null) {
            refreashInfos();
        }
        return info;
    }

    public void unInstall(String id) throws Exception {
        close(id);
        installer.unInstall(appInfos.get(id));
        refreashInfos();
    }

    private WebViewFragment findFragmentById(String id) {
        FragmentManager manager = activity.getSupportFragmentManager();
        List<Fragment> fragments = manager.getFragments();
        for (int i = 0; i < fragments.size(); i++) {
            WebViewFragment fragment = (WebViewFragment)fragments.get(i);
            if ((fragment != null) && (fragment.id.equals(id))) {
                return fragment;
            }
        }
        return null;
    }

    public void switchContent(WebViewFragment fragment, String id) {
        FragmentManager manager = activity.getSupportFragmentManager();
        FragmentTransaction transaction = manager.beginTransaction();
        if ((curFragment != null) && (curFragment != fragment)) {
            transaction.hide(curFragment);
//            if (!curFragment.id.equals("launcher")) {
//                curFragment.onPause();
//            }
        }
        if (curFragment != fragment) {
            if (!fragment.isAdded()) {
                transaction.add(R.id.content, fragment, id);
            } else if (curFragment != fragment) {
                transaction.setCustomAnimations(android.R.animator.fade_in, android.R.animator.fade_out)
                        .show(fragment);
            }
//            transaction.addToBackStack(null);
            transaction.commit();
//            if (!id.equals("launcher")) {
//                fragment.onResume();
//            }
            curFragment = fragment;
        }
        lastList.remove(id);
        lastList.add(0, id);
    }

    public boolean doBackPressed() {
        if (curFragment.id.equals("launcher")) {
            return true;
        }
        else {
            switchContent(findFragmentById("launcher"), "launcher");
            return false;
        }
    }

    public void start(String id) throws Exception {
        WebViewFragment fragment = findFragmentById(id);
        if (fragment == null) {
            if (id.equals("launcher")) {
                fragment = LauncherViewFragment.newInstance();
            }
            else {
                AppInfo info = appInfos.get(id);
                if (info == null) {
                    throw new Exception("No such app!");
                }
                fragment = AppViewFragment.newInstance(id);
            }
        }
//        else if (!id.equals("launcher")) {
//            fragment.onResume();
//        }
        switchContent(fragment, id);
    }

    public void close(String id) throws Exception {
        if (id.equals("launcher")) {
            throw new Exception("Launcher can't close!");
        }

        AppInfo info = appInfos.get(id);
        if (info == null) {
            throw new Exception("No such app!");
        }

        FragmentManager manager = activity.getSupportFragmentManager();
        WebViewFragment fragment = findFragmentById(id);
        if (fragment == null) {
            return;
        }

        if (fragment == curFragment) {
            String id2 = lastList.get(1);
            WebViewFragment fragment2 = findFragmentById(id2);
            if (fragment2 == null) {
                throw new Exception("RT inner error!");
            }
            switchContent(fragment2, id2);
        }

//        if (fragment.appView != null) {
//            fragment.appView.handleDestroy();
//        }

        FragmentTransaction transaction = manager.beginTransaction();
        transaction.remove(fragment);
        transaction.commit();
        lastList.remove(id);
    }

    public void loadLauncher() throws Exception {
        start("launcher");
    }

    public void setInstallUri(String uri) {
        if (uri == null) return;

        if (launcherReady) {
            try {
                sendMessage("launcher", MSG_TYPE_EXTERNAL_INSTALL, uri, "system");
            }
            catch (Exception e) {
                e.printStackTrace();
            }
        }
        else {
            installUriList.add(uri);
        }
    }

    public boolean isLauncherReady() {
        return launcherReady;
    }

    public void setLauncherReady() {
        launcherReady = true;

        for (int i = 0; i < installUriList.size(); i++) {
            String uri = installUriList.get(i);
            try {
                sendMessage("launcher", MSG_TYPE_EXTERNAL_INSTALL, uri, "system");
            }
            catch (Exception e) {
                e.printStackTrace();
            }
        }
    }


    public void sendMessage(String toId, int type, String msg, String fromId) throws Exception {
        FragmentManager manager = activity.getSupportFragmentManager();
        WebViewFragment fragment = (WebViewFragment)manager.findFragmentByTag(toId);
        if (fragment != null) {
            fragment.basePlugin.onReceive(msg, type, fromId);
        }
        else {
            throw new Exception(toId + " isn't running!");
        }
    }

    public int getPluginAuthority(String id, String plugin) {
        AppInfo info = appInfos.get(id);
        if (info != null) {
            for (AppInfo.PluginAuth pluginAuth : info.plugins) {
                if (pluginAuth.plugin.equals(plugin)) {
                    return pluginAuth.authority;
                }
            }
        }
        return AppInfo.AUTHORITY_NOEXIST;
    }

    public int getUrlAuthority(String id, String url) {
        AppInfo info = appInfos.get(id);
        if (info != null) {
            for (AppInfo.UrlAuth urlAuth : info.urls) {
                if (urlAuth.url.equals(url)) {
                    return urlAuth.authority;
                }
            }
        }
        return AppInfo.AUTHORITY_NOEXIST;
    }

    public void setPluginAuthority(String id, String plugin, int authority) throws Exception {
        AppInfo info = appInfos.get(id);
        if (info == null) {
            throw new Exception("No such app!");
        }

        dbAdapter.updatePluginAuth(info.tid, plugin, authority);
        for (AppInfo.PluginAuth pluginAuth : info.plugins) {
            if (pluginAuth.plugin.equals(plugin)) {
                pluginAuth.authority = authority;
                return;
            }
        }
        throw new Exception("The plugin isn't in list!");
    }

    public void setUrlAuthority(String id, String url, int authority)  throws Exception {
        AppInfo info = appInfos.get(id);
        if (info == null) {
            throw new Exception("No such app!");
        }

        int count = dbAdapter.updateURLAuth(info.tid, url, authority);
        if (count > 0) {
            for (AppInfo.UrlAuth urlAuth : info.urls) {
                if (urlAuth.url.equals(url)) {
                    urlAuth.authority = authority;
                    return ;
                }
            }
        }
        throw new Exception("The plugin isn't in list!");
    }

    private static void print(String msg) {
        String name = Thread.currentThread().getName();
        System.out.println(name + ": " + msg);
    }


    private class LockObj {
        int authority = AppInfo.AUTHORITY_NOINIT;
    }

    public synchronized int runAlertPluginAuth(AppInfo info, String plugin, int originAuthority) {
        LockObj lock = new LockObj();
        lock.authority = originAuthority;
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                alertPluginAuth(info, plugin, lock);
            }
        });
        try {
            synchronized (lock) {
                if (lock.authority == originAuthority) {
                    lock.wait();
                }
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
            return originAuthority;
        }
        return lock.authority;
    }

    public void alertPluginAuth(AppInfo info, String plugin, LockObj lock) {
        AlertDialog.Builder ab = new AlertDialog.Builder(activity);
        ab.setTitle("Plugin authority request");
        ab.setMessage("App:'" + info.name + "' request plugin:'" + plugin + "' access authority.");
        ab.setIcon(android.R.drawable.ic_dialog_info);


        ab.setPositiveButton("Allow", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                try {
                    setPluginAuthority(info.app_id, plugin, AppInfo.AUTHORITY_ALLOW);
                }
                catch (Exception e) {
                    e.printStackTrace();
                }
                synchronized (lock) {
                    lock.authority = AppInfo.AUTHORITY_ALLOW;
                    lock.notify();
                }
            }
        });
        ab.setNegativeButton("Refuse", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                try {
                    setPluginAuthority(info.app_id, plugin, AppInfo.AUTHORITY_DENY);
                }
                catch (Exception e) {
                    e.printStackTrace();
                }
                synchronized (lock) {
                    lock.authority = AppInfo.AUTHORITY_DENY;
                    lock.notify();
                }
            }
        });
        ab.show();
    }

    public synchronized int runAlertUrlAuth(AppInfo info, String url, int originAuthority) {
        LockObj lock = new LockObj();
        lock.authority = originAuthority;
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                alertUrlAuth(info, url, lock);
            }
        });
        try {
            synchronized (lock) {
                if (lock.authority == originAuthority) {
                    lock.wait();
                }
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
            return originAuthority;
        }
        return lock.authority;
    }

    public void alertUrlAuth(AppInfo info, String url, LockObj lock) {
        AlertDialog.Builder ab = new AlertDialog.Builder(activity);
        ab.setTitle("Url authority request");
        ab.setMessage("App:'" + info.name + "' request url:'" + url + "' access authority.");
        ab.setIcon(android.R.drawable.ic_dialog_info);

        ab.setPositiveButton("Allow", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                try {
                    setUrlAuthority(info.app_id, url, AppInfo.AUTHORITY_ALLOW);
                }
                catch (Exception e) {
                    e.printStackTrace();
                }
                synchronized (lock) {
                    lock.authority = AppInfo.AUTHORITY_ALLOW;
                    lock.notify();
                }
            }
        });
        ab.setNegativeButton("Refuse", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                try {
                    setUrlAuthority(info.app_id, url, AppInfo.AUTHORITY_DENY);
                }
                catch (Exception e) {
                    e.printStackTrace();
                }
                synchronized (lock) {
                    lock.authority = AppInfo.AUTHORITY_DENY;
                    lock.notify();
                }
            }
        });
        ab.show();
    }

    public String[] getAppList() {
        String[] ids = new String[appList.length];
        for (int i = 0; i < appList.length; i++) {
            ids[i] = appList[i].app_id;
        }
        return ids;
    }

    public String[] getRunningList() {
        FragmentManager manager = activity.getSupportFragmentManager();
        List<Fragment> fragments = manager.getFragments();
        String[] ids = new String[fragments.size()];

        for (int i = 0; i < fragments.size(); i++) {
            WebViewFragment fragment = (WebViewFragment)fragments.get(i);
            ids[i] = fragment.id;
        }
        return ids;
    }

    public String[] getLastList() {
        String[] ids = new String[lastList.size()];
        return lastList.toArray(ids);
    }


    public void flingTheme() {
        if (curFragment.id.equals("launcher")) {
            return;
        }

        AppViewFragment fragment = (AppViewFragment) curFragment;
        if (fragment.titlebar.getVisibility() == View.VISIBLE) {
            fragment.titlebar.setVisibility(View.GONE);
        } else {
            fragment.titlebar.bringToFront();//for qrscanner
            fragment.titlebar.setVisibility(View.VISIBLE);
        }
    }

    public void onConfigurationChanged(Configuration newConfig) {
        FragmentManager manager = activity.getSupportFragmentManager();
        List<Fragment> fragments = manager.getFragments();

        for (int i = 0; i < fragments.size(); i++) {
            WebViewFragment fragment = (WebViewFragment)fragments.get(i);
            if (fragment != null && fragment.appView != null) {
                PluginManager pm = fragment.appView.getPluginManager();
                if (pm != null) {
                    pm.onConfigurationChanged(newConfig);
                }
            }
        }
    }

    public void onRequestPermissionResult(int requestCode, String permissions[],
                                           int[] grantResults) throws JSONException {
        FragmentManager manager = activity.getSupportFragmentManager();
        List<Fragment> fragments = manager.getFragments();

        for (int i = 0; i < fragments.size(); i++) {
            WebViewFragment fragment = (WebViewFragment)fragments.get(i);
            if (fragment != null) {
                fragment.onRequestPermissionResult(requestCode, permissions, grantResults);
            }
        }
    }
}

File: platform_src/ios/Elastos/Classes/AppManager.swift
 /*
  * Copyright (c) 2018 Elastos Foundation
  *
  * Permission is hereby granted, free of charge, to any person obtaining a copy
  * of this software and associated documentation files (the "Software"), to deal
  * in the Software without restriction, including without limitation the rights
  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  * copies of the Software, and to permit persons to whom the Software is
  * furnished to do so, subject to the following conditions:
  *
  * The above copyright notice and this permission notice shall be included in all
  * copies or substantial portions of the Software.
  *
  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */

import Foundation

class AppManager {
    private static var appManager: AppManager?;
    
    /** The internal message */
    static let MSG_TYPE_INTERNAL = 1;
    /** The internal return message. */
    static let MSG_TYPE_IN_RETURN = 2;
    /** The external launcher message */
    static let MSG_TYPE_EXTERNAL_LAUNCHER = 3;
    /** The external install message */
    static let MSG_TYPE_EXTERNAL_INSTALL = 4;
    /** The external return message. */
    static let MSG_TYPE_EX_RETURN = 5;
    
    let mainViewController: MainViewController;
    var viewControllers = [String: TrinityViewController]();
    
    let appsPath: String;
    let dataPath: String;
    let tempPath: String;
    
    var curController: TrinityViewController?;
    
    let dbAdapter: ManagerDBAdapter;
    
    var appList: [AppInfo];
    var appInfos = [String: AppInfo]();
    var lastList = [String]();
    let installer: AppInstaller;
    
    private var launcherInfo: AppInfo? = nil;
    
    var installUriList = [String]();
    var launcherReady = false;
    
    init(_ mainViewController: MainViewController) {
        self.mainViewController = mainViewController;
        appsPath = NSHomeDirectory() + "/Documents/apps/";
        dataPath = NSHomeDirectory() + "/Documents/data/";
        tempPath = NSHomeDirectory() + "/Documents/temp/";
        
        let fileManager = FileManager.default
        var first = false;
        if (!fileManager.fileExists(atPath: appsPath)) {
            do {
                try fileManager.createDirectory(atPath: appsPath, withIntermediateDirectories: true, attributes: nil)
                first = true;
            }
            catch let error {
                print("Make appsPath error: \(error)");
            }     
        }
        
        if (!fileManager.fileExists(atPath: dataPath)) {
            do {
                try fileManager.createDirectory(atPath: dataPath, withIntermediateDirectories: true, attributes: nil)
            }
            catch let error {
                print("Make dataPath error: \(error)");
            }
        }
        
        if (!fileManager.fileExists(atPath: tempPath)) {
            do {
                try fileManager.createDirectory(atPath: tempPath, withIntermediateDirectories: true, attributes: nil)
            }
            catch let error {
                print("Make tempPath error: \(error)");
            }
        }
        
        dbAdapter = ManagerDBAdapter(dataPath);
        installer = AppInstaller(appsPath, dataPath, tempPath, dbAdapter);
        appList = try! dbAdapter.getAppInfos();
        
        if first {
            saveLauncherInfo();
        }
        
        saveBuiltInAppInfos();
        refreashInfos();
        
        AppManager.appManager = self;
    }
    
    static func getShareInstance() -> AppManager {
        return AppManager.appManager!;
    }
    
    func saveLauncherInfo() {
        let path = getAbsolutePath("www/launcher");
        do {
            let info = try installer.parseManifest(path + "/manifest.json", true)!;
            try installer.copyAssetsFolder(path, appsPath + info.app_id);
            info.built_in = true;
            try dbAdapter.addAppInfo(info);
        }
        catch let error {
            print("Copy launcher error: \(error)");
        }
    }
    
    func getLauncherInfo() -> AppInfo {
        if launcherInfo == nil {
            launcherInfo = try! dbAdapter.getLauncherInfo();
        }
        return launcherInfo!;
    }
    
    func refreashInfos() {
        appList = try! dbAdapter.getAppInfos();
        appInfos = [String: AppInfo]();
        for info in appList {
            appInfos[info.app_id] = info;
        }
    }
    
    func getAppInfo(_ id: String) -> AppInfo? {
        if (id == "launcher") {
            return getLauncherInfo();
        }
        else {
            return appInfos[id];
        }
    }
    
    func getAppInfos() -> [String: AppInfo] {
        return appInfos;
    }
    
    func getStartPath(_ info: AppInfo) -> String {
        if (!info.remote) {
            return getAppUrl(info) + info.start_url;
        }
        else {
            return info.start_url;
        }
    }
    
    func getAppPath(_ info: AppInfo) -> String {
        if (!info.remote) {
            return appsPath + info.app_id + "/";
        }
        else {
            let index = info.start_url.range(of: "/", options: .backwards)?.lowerBound
            return String(info.start_url[info.start_url.startIndex ..< index!]);
        }
    }
    
    func getAppUrl(_ info: AppInfo) -> String {
        var url = getAppPath(info);
        if (!info.remote) {
            url = "file://" + url;
        }
        return url;
    }
    
    func getDataPath(_ id: String) -> String {
        var appId = id;
        if (id == "launcher") {
            appId = getLauncherInfo().app_id;
        }
        return dataPath + appId + "/";
    }
    
    func getDataUrl(_ id: String) -> String {
        return "file://" + getDataPath(id);
    }
    
    func getTempPath(_ id: String) -> String {
        var appId = id;
        if (id == "launcher") {
            appId = getLauncherInfo().app_id;
        }
        return tempPath + appId + "/";
    }
    
    func getTempUrl(_ id: String) -> String {
        return "file://" + getTempPath(id);
    }
    
    func saveBuiltInAppInfos() {
        let path = getAbsolutePath("www/built-in") + "/";
        
        let fileManager = FileManager.default;
        let dirs = try? fileManager.contentsOfDirectory(atPath: path);
        guard dirs != nil else {
            return;
        }
        
        do {
            for dir in dirs! {
                var needInstall = true;
                for info in appList {
                    if info.app_id == dir {
                        needInstall = false;
                        break;
                    }
                }
                
                if (needInstall) {
                    try installer.copyAssetsFolder(path +  dir, appsPath + dir);
                    let info = try installer.parseManifest(appsPath +  dir + "/manifest.json")
                    guard (info != nil || info!.app_id != "") else {
                        return;
                    }
                    
                    info!.built_in = true;
                    try dbAdapter.addAppInfo(info!);
                }
            }
        } catch AppError.error(let err) {
            print(err);
        } catch let error {
            print(error.localizedDescription);
        }
    }

    func install(_ url: String) throws -> AppInfo? {
        let info = try! installer.install(self, url);
        if (info != nil) {
            refreashInfos();
        }
        return info;
    }
    
    func unInstall(_ id: String) throws {
        try close(id);
        try installer.unInstall(appInfos[id]);
        refreashInfos();
    }
    
    func removeLastlistItem(_ id: String) {
        for (index, item) in lastList.enumerated() {
            if item == id {
                lastList.remove(at: index);
                break;
            }
        }
    }
    
    func switchContent(_ to: TrinityViewController, _ id: String) {
        mainViewController.switchController(from: curController!, to: to)
        removeLastlistItem(id);
        lastList.insert(id, at: 0);
    }

    func start(_ id: String) throws {
        var viewController = viewControllers[id]
        if viewController == nil {
            if (id == "launcher") {
                viewController = LauncherViewController();
            }
            else {
                let appInfo = appInfos[id];
                guard appInfo != nil else {
                    throw AppError.error("No such app!");
                }
                let appViewController = AppViewController();
                appViewController.setInfo(id, appInfo!);
                viewController = appViewController;
//                viewController?.startPage = "built-in/org.elastos.trinity.demo1/demo1.html";
            }
            
            mainViewController.add(viewController!)
            viewControllers[id] = viewController;
            lastList.insert(id, at: 0);
        }
        else {
            if (curController != viewController) {
                switchContent(viewController!, id);
            }
        }
        
        curController = viewController
    }
    
    func close(_ id: String) throws {
        if (id == "launcher") {
            throw AppError.error("Launcher can't close!");
        }
        
        let info = appInfos[id];
        if (info == nil) {
            throw AppError.error("No such app!");
        }
    
        let viewController = viewControllers[id]
        if (viewController == nil) {
            return;
        }
    
        if (viewController == curController) {
            let id2 = lastList[1];
            let viewController2 = viewControllers[id2]
            if (viewController2 == nil) {
                throw AppError.error("RT inner error!");
            }
            switchContent(viewController2!, id2);
        }
        
        removeLastlistItem(id);
        viewControllers[id] = nil;
        viewController!.remove();
    }
    
    
    func loadLauncher() throws {
        try start("launcher");
    }
    
    func setInstallUri(_ uri: String) {
        if launcherReady {
            try? self.sendMessage("launcher", AppManager.MSG_TYPE_EXTERNAL_INSTALL, uri, "system");
        }
        else {
            installUriList.append(uri);
        }
    }
    
    func isLauncherReady() -> Bool {
        return launcherReady;
    }
    
    func setLauncherReady() {
        launcherReady = true;
    
        for uri in installUriList {
            try? self.sendMessage("launcher", AppManager.MSG_TYPE_EXTERNAL_INSTALL, uri, "system");
        }
    }

    func sendMessage(_ toId: String, _ type: Int, _ msg: String, _ fromId: String) throws {
        let viewController = viewControllers[toId]
        if (viewController != nil) {
            viewController!.basePlugin!.onReceive(msg, type, fromId);
        }
        else {
            throw AppError.error(toId + " isn't running!");
        }
    }
    
    func getPluginAuthority(_ id: String, _ plugin: String) -> Int {
        let info = appInfos[id];
        if (info != nil) {
            for pluginAuth in info!.plugins {
                if (pluginAuth.plugin == plugin) {
                    return pluginAuth.authority;
                }
            }
        }
        return AppInfo.AUTHORITY_NOEXIST;
    }
    
    func getUrlAuthority(_ id: String, _ url: String) -> Int {
        let info = appInfos[id];
        if (info != nil) {
            for urlAuth in info!.urls {
                if (urlAuth.url == url) {
                    return urlAuth.authority;
                }
            }
        }
        return AppInfo.AUTHORITY_NOEXIST;
    }
    
    func setPluginAuthority(_ id: String, _ plugin: String, _ authority: Int) throws {
        let info = appInfos[id];
        guard (info != nil) else {
            throw AppError.error("No such app!");
        }
        
        for pluginAuth in info!.plugins {
            if (pluginAuth.plugin == plugin) {
                try dbAdapter.updatePluginAuth(pluginAuth, authority);
                pluginAuth.authority = authority;
                return;
            }
        }
        throw AppError.error("The plugin isn't in list!");
    }
    
    func setUrlAuthority(_ id: String, _ url: String, _ authority: Int) throws {
        let info = appInfos[id];
        guard (info != nil) else {
            throw AppError.error("No such app!");
        }
        
        for urlAuth in info!.urls {
            if (urlAuth.url == url) {
                try dbAdapter.updateUrlAuth(urlAuth, authority);
                urlAuth.authority = authority;
                return;
            }
        }
        throw AppError.error("The url isn't in list!");
    }

    func runAlertPluginAuth(_ info: AppInfo, _ pluginName: String,
                            _ plugin: TrinityPlugin,
                            _ command: CDVInvokedUrlCommand) {

        func doAllowHandler(alerAction:UIAlertAction) {
            try? setPluginAuthority(info.app_id, pluginName, AppInfo.AUTHORITY_ALLOW);
            plugin.execute(command);
            let result = CDVPluginResult(status: CDVCommandStatus_NO_RESULT);
            result?.setKeepCallbackAs(false);
            plugin.commandDelegate?.send(result, callbackId:command.callbackId);
        }
        
        func doRefuseHandler(alerAction:UIAlertAction) {
            try? setPluginAuthority(info.app_id, pluginName, AppInfo.AUTHORITY_DENY);
            let result = CDVPluginResult(status: CDVCommandStatus_ERROR,
                                         messageAs: "Plugin:'" + pluginName + "' have not run authority.");
            result?.setKeepCallbackAs(false);
            plugin.commandDelegate.send(result, callbackId: command.callbackId)
        }
        
        let alertController = UIAlertController(title: "Plugin authority request",
                message: "App:'" + info.name + "' request plugin:'" + pluginName + "' access authority.",
                preferredStyle: UIAlertController.Style.alert)
        let cancelAlertAction = UIAlertAction(title: "Refuse", style: UIAlertAction.Style.cancel, handler: doRefuseHandler)
        alertController.addAction(cancelAlertAction)
        let sureAlertAction = UIAlertAction(title: "Allow", style: UIAlertAction.Style.default, handler: doAllowHandler)
        alertController.addAction(sureAlertAction)
        
        self.mainViewController.present(alertController, animated: true, completion: nil)
    }
    
    func runAlertUrlAuth(_ info: AppInfo, _ url: String) {
        func doAllowHandler(alerAction:UIAlertAction) {
            try? setUrlAuthority(info.app_id, url, AppInfo.AUTHORITY_ALLOW);
        }
        
        func doRefuseHandler(alerAction:UIAlertAction) {
            try? setUrlAuthority(info.app_id, url, AppInfo.AUTHORITY_DENY);
        }
        
        let alertController = UIAlertController(title: "Url authority request",
                                                message: "App:'" + info.name + "' request url:'" + url + "' access authority.",
                                                preferredStyle: UIAlertController.Style.alert)
        let cancelAlertAction = UIAlertAction(title: "Refuse", style: UIAlertAction.Style.cancel, handler: doRefuseHandler)
        alertController.addAction(cancelAlertAction)
        let sureAlertAction = UIAlertAction(title: "Allow", style: UIAlertAction.Style.default, handler: doAllowHandler)
        alertController.addAction(sureAlertAction)
        self.mainViewController.present(alertController, animated: true, completion: nil)
    }
    
    func getRunningList() -> [String] {
        var ret = [String]();
        for id in viewControllers.keys {
            ret.append(id);
        }
        return ret;
    }
    
    func getAppList() -> [String] {
        var ret = [String]();
        for info in appList {
            ret.append(info.app_id);
        }
        return ret;
    }
    
    func getLastList() -> [String] {
        return lastList;
    }
    
}
